plenum/server/node.py

```diff
@@ -1444,16 +1522,81 @@ def postRecvTxnFromCatchup(self, ledgerId: int, txn: Any):
+
+        # TODO: Maybe a slight optimisation is to check result of
+        # `self.num_txns_caught_up_in_last_catchup()`
+        self.processStashedOrderedReqs()
+
+        if self.is_catchup_needed():
+            logger.debug('{} needs to catchup again'.format(self))
+            self.start_catchup()
+        else:
+            logger.debug('{} does not need any more catchups'.format(self))
+            self.no_more_catchups_needed()
+
+    def is_catchup_needed(self) -> bool:
+        """
+        Check if received a quorum of view change done messages and if yes
+        check if caught up till the
+        Check if all requests ordered till last prepared certificate
+        Check if last catchup resulted in no txns
+        """
+        if self.caught_up_for_current_view():
+            logger.info('{} is caught up for the current view {}'.format(self, self.viewNo))
+            return False
+        logger.debug('{} is not caught up for the current view {}'.format(self, self.viewNo))
+        if self.has_ordered_till_last_prepared_certificate() and \
+                        self.num_txns_caught_up_in_last_catchup() == 0:
+            return False
 
+        return True
+
+    def caught_up_for_current_view(self):
+        if not self.elector._hasViewChangeQuorum(0):
+            logger.debug('{} does not have view change quorum for view {}'.
+                         format(self, self.viewNo))
+            return False
+        vc = self.elector.has_sufficient_same_view_change_done_messages(0)
+        if not vc:
+            logger.debug('{} does not have acceptable ViewChangeDone for '
+                         'view {}'.format(self, self.viewNo))
+            return False
+        ledger_info = vc[1]
+        for lid, size, root_hash in ledger_info:
+            ledger = self.ledgerManager.ledgerRegistry[lid].ledger
+            if size == 0:
+                continue
+            if ledger.size < size:
+                return False
+            if ledger.hashToStr(ledger.tree.merkle_tree_hash(0, size)) != root_hash:
+                return False
+        return True
+
+    def has_ordered_till_last_prepared_certificate(self):
+        lst = self.master_replica.last_prepared_before_view_change
+        if lst is None:
+            return True
+        return compare_3PC_keys(lst, self.master_replica.last_ordered_3pc) >= 0
+
+    def num_txns_caught_up_in_last_catchup(self):
+        count = sum([l.num_txns_caught_up for l in
+                    self.ledgerManager.ledgerRegistry.values()])
+        logger.debug('{} caught up to {} txns in the last catchup'.format(self, count))
+        return count
+
+    def no_more_catchups_needed(self):
+        # This method is called when no more catchups needed
         self.mode = Mode.participating
-        self.processStashedOrderedReqs()
-        # TODO: next line not needed
-        self.checkInstances()
+        # self.processStashedOrderedReqs()
+        self.decidePrimaries()
```

Comment:
We can get here on the second condition for `is_catchup_needed`, that is when we still doesn't have a quorum of ViewChangeDone msgs.
`self.decidePrimaries()` eventually makes node participating and sets a new primary.
So, we may set a new primary and make node participating when we're not ready yet, isn't it? This looks incorrect.